crates/fuel-core/src/service/adapters/graphql_api/off_chain.rs

```diff
@@ -233,15 +233,49 @@ impl OffChainDatabase for OffChainIterableKeyValueView {
         }
     }
 
-    fn balances(
-        &self,
+    fn balances<'a>(
+        &'a self,
         owner: &Address,
-        base_asset_id: &AssetId,
+        start: Option<AssetId>,
+        base_asset_id: &'a AssetId,
         direction: IterDirection,
-    ) -> BoxedIter<'_, StorageResult<(AssetId, TotalBalanceAmount)>> {
+    ) -> BoxedIter<'a, StorageResult<(AssetId, TotalBalanceAmount)>> {
+        match (direction, start) {
+            (IterDirection::Forward, None) => {
+                self.base_asset_first(owner, base_asset_id, direction)
+            }
+            (IterDirection::Forward, Some(asset_id)) => {
+                if asset_id == *base_asset_id {
+                    self.base_asset_first(owner, base_asset_id, direction)
+                } else {
+                    let start = (asset_id != *base_asset_id)
+                        .then_some(CoinBalancesKey::new(owner, &asset_id));
+                    self.non_base_asset_balances(owner, start, direction, base_asset_id)
+                }
+            }
+            (IterDirection::Reverse, None) => {
+                self.non_base_assets_first(start, owner, base_asset_id, direction)
+            }
+            (IterDirection::Reverse, Some(asset_id)) => {
+                if asset_id == *base_asset_id {
```

Comment:
You can leverage pattern matching as you did in `non_base_asset_balances` don't really know if it's better or not. up to you :)

Reply:
I think I'll leave it as is, should be ok.